validate properties on init

EndangeredMassa opened the pull request:
This change validates properties on `init` of the component.

It will NOT re-validate them if they change after `init`, currently. This probably should be added, just in case, though. Unfortunately, we can't use computed properties for `tagName` because Ember prevents that. Should we set up an observer?

---

Does this meet the requirements you mentioned, @MelSumner?

EndangeredMassa committed:
validate properties on init

MelSumner commented:
- [ ] does Ember already prevent us from trying to change tagName after init? If not then this _should_ break if the user tries to do that. 
I don't have the answer for that this second, and I want to try to test the edges, but it seems as though this satisfies the concerns we discussed. I will revisit this tomorrow. 👍

MelSumner commented on addon/components/a11y-landmark.js:
```diff
@@ -12,6 +12,28 @@ const LANDMARK_NAVIGATION_ROLE = {
+const VALID_LANDMARK_ROLES = [
+    'banner',
+    'navigation',
+    'aside',
```
the aside role is "complementary"

MelSumner replied:
@EndangeredMassa to clarify here- an "aside" is not a valid landmark role. It would need to be 'complementary' instead of 'aside' under VALID_LANDMARK_ROLES.

MelSumner commented on addon/components/a11y-landmark.js:
```diff
@@ -12,6 +12,28 @@ const LANDMARK_NAVIGATION_ROLE = {
+    'search',
+    'application',
+    'document',
+    'region'
```
'contentinfo' role is missing.

MelSumner commented on addon/components/a11y-landmark.js:
```diff
@@ -60,5 +87,22 @@ export default Ember.Component.extend({
+            Ember.assert(`Invalid tagName "${landmarkRole}". Must be one of ${validValues}.`);
+        }
+    }
+});
```
So this part needs a bit of tweaking. 

For example, if I put this in my template:

`{{#a11y-landmark tagName="header" landmarkRole="contentInfo"}}
    <h1>Header</h1>
{{/a11y-landmark}}
`
This returns the error 
> Invalid tagName "contentInfo". Must be one of banner, navigation, aside, main, form, search, application, document, region.

What it should be doing could be one of two things:
1. if there's a tagName, declared (it means they meant it to be something other than div), ignore the landmarkRole, since it adds it for them. 
2. however, we could still assert that they are wrong (a valid tagName that isn't div was declared, you don't need to assign a landmarkRole) to help users learn how to do this correctly.

ksin commented on addon/components/a11y-landmark.js:
```diff
@@ -60,5 +87,22 @@ export default Ember.Component.extend({
 
-});
\ No newline at end of file
+    _validateTagName(tagName) {
```
can probably DRY this up. something like:
```js
_validate(attribute, list) {
  if (list.indexOf(attribute) === -1) { return; }
  const validValues = list.join(', ');
  Ember.assert(`Invalid tagName "${attribute}". Must be one of ${validValues}.`);
} 
```

If you want, you can make the lists a string in the format of `validValues` since you're going to join them every time anyway. Not a significant performance improvement, but you can see the effects: https://jsperf.com/stringarrayindexof

EndangeredMassa replied:
Yup! We can certainly do DRY things up if they stay in this form. I wanted to get feedback first and refactor things like this afterwards. It's more work if we DRY things up, then have to pull them back apart.

EndangeredMassa committed:
refactor and test the permutations of properties that lead to ariaRoles

EndangeredMassa commented:
Thanks for the clarification!

The latest commit does the following: 

- The consumer must specify one or the other, but not neither or both, of `tagName` and `landmarkRole`. (There is an exception made for `tagName="form" landmarkRole="search"` to result in `ariaRole` being `search`.)
- The validation happens inside the `ariaRole` computed property, which will be executed on render. This will now recalculate properly and better encapsulates the behavior for determining `ariaRole`.
- The `ariaRole` functionality should now have all branches tested, but not every single permutation of inputs to outputs.

MelSumner commented:
@EndangeredMassa this looks pretty great, just a few tweaks for accuracy and I'll merge the PR.

EndangeredMassa committed:
change a valid landmark role from aside to complementary

EndangeredMassa committed:
add contentinfo role to valid landmark roles

EndangeredMassa commented:
- switched aside/complementary
- added contentinfo role

Is `document` a landmark role? In the spec, it's listed under [document structure roles](https://www.w3.org/TR/wai-aria-1.1/#document_structure_roles), but I could be missing something.

MelSumner commented:
@EndangeredMassa document is a better default landmark role now that I think about it more deeply. The body is a document role by default. You'd use document as the role when you wanted the native browser behavior to resume inside of an element that was given the role of application. It ends up being, in my experience, a super edge case (since the use of role="application" itself is supposed to be incredibly sparse).
